clang/lib/CodeGen/CGClass.cpp

```diff
@@ -2988,8 +2989,9 @@ void CodeGenFunction::EmitVTablePtrCheck(const CXXRecordDecl *RD,
 }
 
 bool CodeGenFunction::ShouldEmitVTableTypeCheckedLoad(const CXXRecordDecl *RD) {
-  if (!CGM.getCodeGenOpts().WholeProgramVTables ||
-      !CGM.HasHiddenLTOVisibility(RD))
+  if ((!CGM.getCodeGenOpts().WholeProgramVTables ||
+       !CGM.HasHiddenLTOVisibility(RD)) &&
+      !CGM.getCodeGenOpts().DevirtualizeSpeculatively)
```

Comment:
We seem to do that check in multiple places - maybe a `ShouldPerformSpeculativeDevirtualization()` function would be useful

Reply:
It might be nice to refactor the whole if statement, although there are a few different variations of the combination of conditions being checked. But yes, it should be named something more general, like shouldEmitVTableTypeInfo or something like that.